xdsl/transforms/csl_wrapper_to_csl.py

```diff
@@ -0,0 +1,242 @@
+from xdsl.context import MLContext
+from xdsl.dialects import arith, builtin, scf
+from xdsl.dialects.csl import csl, csl_wrapper
+from xdsl.ir import Block, Region, SSAValue
+from xdsl.irdl import base
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+from xdsl.utils.hints import isa
+from xdsl.utils.isattr import isattr
+
+
+def _get_prog_name(
+    from_wrapper: builtin.StringAttr | None, from_transofrm: str | None
+) -> str:
+    if from_wrapper is not None:
+        return from_wrapper.data
+    if from_transofrm is not None:
+        return from_transofrm
+    return "pe_program"
+
+
+def _collect_params(op: csl_wrapper.ModuleOp):
+    new_args = list[SSAValue]()
+    for param in op.params:
+        if isa(param.value, builtin.IntegerAttr):
+            value = arith.Constant(param.value)
+        else:
+            value = None
+        p = csl.ParamOp(param.key.data, param.type, value)
+        new_args.append(SSAValue.get(p))
+    return new_args
+
+
+def _add_to_toplevel(
+    rewriter: PatternRewriter, op: csl_wrapper.ModuleOp, mod: csl.CslModuleOp
+):
+    assert isa(builtin_module := op.parent_op(), builtin.ModuleOp)
+    rewriter.insert_op(mod, InsertPoint.at_start(builtin_module.regions[0].block))
+
+
+@dataclass(frozen=True)
+class RemoveWrapper(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, _: csl_wrapper.ModuleOp, rewriter: PatternRewriter, /):
+        rewriter.erase_matched_op()
+
+
+@dataclass(frozen=True)
+class ExtractLayoutModule(RewritePattern):
+    prog_name: str | None = None
+
+    def add_tile_code(
+        self,
+        x: SSAValue,
+        y: SSAValue,
+        width: csl.ParamOp,
+        height: csl.ParamOp,
+        yield_op: csl_wrapper.YieldOp,
+        prog_name: str,
+    ):
+        struct = csl.ConstStructOp(
+            ("width", width),
+            ("height", height),
+            *(f for f in yield_op.items()),
+        )
+        return (
+            struct,
+            csl.SetTileCodeOp(fname=prog_name, x_coord=x, y_coord=y, params=struct),
+        )
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: csl_wrapper.ModuleOp, rewriter: PatternRewriter, /):
+        prog_name = _get_prog_name(op.program_name, self.prog_name)
+        module_block = Block()
+
+        outer_loop_block = Block()
+        outer_loop_block.insert_arg(builtin.IntegerType(16), 0)
+        x = outer_loop_block.args[0]
+
+        inner_loop_block = Block()
+        inner_loop_block.insert_arg(builtin.IntegerType(16), 0)
+        y = inner_loop_block.args[0]
+
+        new_args = list[SSAValue]()
+
+        assert isa(yield_op := op.layout_module.block.last_op, csl_wrapper.YieldOp)
+        rewriter.erase_op(yield_op)
+
+        with ImplicitBuilder(module_block):
```

Comment:
```suggestion
        with ImplicitBuilder(module_block := Block()):
```

Some of these can be inlined, also inner/outer loop block. The block arg types can be passed directly as params.